### PR TITLE
[ILASM] Added `/DET` flag for deterministic compilations

### LINKED ISSUE #8293: [ILASM] Add option to produce deterministic outputs
Various .NET compilers started adding `/deterministic` command line option that makes them produce deterministic output (multiple invocations of the compiler produce byte-for-byte identical outputs for the same input).

This includes fields such as:
* The timestamp hardcoded in the PE header
* The MVID value
* PDB guid
* Full paths to the source files
* Maybe more

### LINKED ISSUE #62484: Symbols generated by ILAsm do not contain a PdbChecksum debug directory entry
The  System.Runtime.CompilerServices.Unsafe 6.0 package does not have symbols on MSDL or NuGet's symbol server. 

@ericstj

### PR DESCRIPTION
**Pull request by TIHan**
## Description
Resolves https://github.com/dotnet/runtime/issues/8293
Resolves https://github.com/dotnet/runtime/issues/62484

Part of this is from the work I did two years ago on the internal version of ILASM for .NET Framework.

Additional changes:
- PdbChecksum and Deterministic Debug Directory entries
- Deterministic Pdbs

## Acceptance Criteria
- [x] SHA256 hash for non-win32
- [x] Investigate Portable PDB determinism (PR includes generating a PDB GUID based on the PDB contents)
- [x] Investigate Full paths to the source files (This is already handled)
- [x] Deterministic tests
    - When running ilasmroundtrip tests, they will also verify deterministic assemblies and pdbs emitted by ILASM.

### CODE REVIEW

**Comment by TIHan** [comment]
@GrabYourPitchforks I need to use a SHA256 hash algo for all platforms, do you have any recommendations?

File: src/coreclr/md/inc/portablepdbmdi.h [hunk 1]
```diff
@@ -69,6 +69,13 @@ DECLARE_INTERFACE_(IMetaDataEmit3, IMetaDataEmit2)
         USHORT      index,                  // [IN] Variable index (slot).
         char        *name,                  // [IN] Variable name.
         mdLocalVariable *locVarToken) PURE; // [OUT] Token of the defined variable.
+
```
Comment (TIHan) [inline src/coreclr/md/inc/portablepdbmdi.h:72]: I have this feeling that I may need to introduce `IMetaDataEmit4`...

Comment (TIHan) [reply src/coreclr/md/inc/portablepdbmdi.h:72]: @davidwrighton is that correct? Even though `IMetaDataEmit3` is only used for the portable PDB feature, if I want to add additional APIs, I will want to create a new one, i.e. `IMetaDataEmit4` ?

Comment (davidwrighton) [reply src/coreclr/md/inc/portablepdbmdi.h:72]: Sadly, yes. Although if this is only for Ilasm... Give a different name.

**Comment by TIHan** [comment]
Re-opening as I want to finish this up.

File: src/coreclr/md/inc/portablepdbmdi.h (hunk 1 above)
Comment (TIHan) [reply src/coreclr/md/inc/portablepdbmdi.h:72]: Ok, named it `IILAsmPortablePdbWriter`.

**Comment by TIHan** [comment]
So far, I'm happy with the changes, but CI times out as I think there is a problem with how I'm constructing the debug directories.

File: src/coreclr/md/enc/sha256.cpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//*****************************************************************************
+// sha256.cpp
```
Comment (jkotas) [inline src/coreclr/md/enc/sha256.cpp:4]: Having our own implementation of sha256 is going to be compliance burden. You should discuss this with the security folks (email `.NET Security Team`).

Comment (TIHan) [reply src/coreclr/md/enc/sha256.cpp:4]: That's fair. Will do.

Before, I had the impl using the built-in bcrypt for windows, but I do not know an equivalent for the other OSes, other than OpenSSL. I think using OpenSSL might be overkill for this, but is there any other alternative?

**Comment by clairernovotny** [comment]
What is the current status of this feature; was it implemented in a diff PR?

**Comment by jkotas** [comment]
This feature is tracked by https://github.com/dotnet/runtime/issues/8293

https://github.com/dotnet/runtime/pull/109091 is an open PR with the implementation.

**Pull request closed without merging**